torch/optim/sradam.py

```diff
@@ -0,0 +1,115 @@
+import copy
+
+import torch
+from .adam import Adam
+
+
+def _apply_square_to_state_dict(state_dict):
+    with torch.no_grad():
+        for state_per_param in state_dict['state'].values():
+            state_per_param['exp_avg_sq'].square_()
+            state_per_param['max_exp_avg_sq'].square()
+    return state_dict
+
+
+def _apply_sqrt_to_state_dict(state_dict):
+    with torch.no_grad():
+        for state_per_param in state_dict['state'].values():
+            state_per_param['exp_avg_sq'].sqrt_()
+            if 'max_exp_avg_sq' not in state_per_param:
+                state_per_param['max_exp_avg_sq'] = torch.zeros_like(state_per_param['exp_avg_sq'])
+            else:
+                state_per_param['max_exp_avg_sq'].sqrt_()
+    return state_dict
+
+
+class SRAdam(Adam):
+    r"""Implements Adam algorithm with Stochastic Rounding.
+
+    It has been proposed in `Adam: A Method for Stochastic Optimization`_.
+
+    With Stochastic Rounding, param, `exp_avg`, `exp_avg_sq`, and optionally `max_exp_avg_sq`
+    can be represented with 16 bits.
+    This optimizer requires CUDA.
+
+    Arguments:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float, optional): learning rate (default: 1e-3)
+        betas (Tuple[float, float], optional): coefficients used for computing
+            running averages of gradient and its square (default: (0.9, 0.999))
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        amsgrad (boolean, optional): whether to use the AMSGrad variant of this
+            algorithm from the paper `On the Convergence of Adam and Beyond`_
+            (default: False)
+
+    .. _Adam\: A Method for Stochastic Optimization:
```

Comment:
Where does the idea of stochastic rounding come from? Is there a paper to cite?

Reply:
Put a link in `torch.stochastic_rounding` docstring and now optimizers refer to it.
https://github.com/csarofeen/pytorch/pull/17/files#diff-cfb065e43691d37ba30f3856fe4352b3R7536